ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,312 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2019, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <boost/tokenizer.hpp>
+#include <cmath>
+
+#include "BaseLib/ConfigTreeUtil.h"
+#include "Output.h"
+#include "PhreeqcInterface.h"
+#include "ThirdParty/iphreeqc/src/IPhreeqc.h"
+
+namespace ChemistryLib
+{
+void phreeqcIO::doWaterChemistryCalculation(
+    std::vector<GlobalVector*>& process_solutions,
+    ProcessLib::Process const& process, double const dt)
+{
+    setAqueousSolutionsOrUpdateProcessSolutions(
+        process_solutions, process, Status::SettingAqueousSolutions);
+    setTimeStep(dt);
+
+    writeInputsToFile();
+
+    execute();
+
+    readOutputsFromFile();
+
+    setAqueousSolutionsOrUpdateProcessSolutions(
+        process_solutions, process, Status::UpdatingProcessSolutions);
+}
+
+void phreeqcIO::setAqueousSolutionsOrUpdateProcessSolutions(
+    std::vector<GlobalVector*> const& process_solutions,
+    ProcessLib::Process const& process, Status const status)
+{
+    int const nNodes = _solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        int process_id = 0;
+        int const variable_id = 0;
+        auto& solution = _solutions[node_id];
+        auto& components = solution._components;
+        for (auto process_solution_iter = process_solutions.begin();
+             process_solution_iter != process_solutions.end();
+             ++process_solution_iter)
```

Comment:
One could use a for-range loop here, or reuse the process_id, which is counted anyway.

Reply:
This part has been rewritten.